### DISCUSSION TITLE
Context swapping with multiple middlewares

### DISCUSSION THREAD

**Discussion by gino** [trpc/trpc#2770] [Q&A] [answered]
Hi there,

I am currently using middleware context swapping for authentication and this works great so far. However, I was wondering if it would be possible to have the expected context swapping work through multiple middlewares and if this is even possible.

I expected the following code to work but unfortunately, it doesn't. I might be doing it wrong but not quite sure:

```ts
t.procedure
  .use(
    t.middleware(({ next }) => {
      return next({ ctx: { user: "John Doe" } });
    })
  )
  .use(
    t.middleware(({ next, ctx }) => {
      const user = ctx.user; // Type error: undefined, even though it is not.
      return next();
    })
  )
  .query(({ ctx }) => {
    return ctx.user; // Works as expected
  });
```

My use case for this is to have the regular authentication check as the first middleware and a permission check as the second middleware, which requires the authenticated user.

This might not even be possible as I didn't see it in the docs but I could be missing something.

Thanks a lot.

**Comment by KATT** [comment, marked as answer]
It should be possible, but you're using the middlewares wrong.

```ts
t.procedure
  .use(
    ({ next }) => {
      return next({ ctx: { user: "John Doe" } });
    }
  )
  .use(
    ({ next, ctx }) => {
      const user = ctx.user;
      return next();
    }
  )
  .query(({ ctx }) => {
    return ctx.user; 
  });
```

**Reply by KATT** [reply]
See https://trpc.io/docs/procedures#multiple-input-parsers on how you can combine this to make a reusable procedure that you can use to check that

1. User is authorized
1. It's a "room procedure"
1. The user has access to this specific "room procedure"
1. Add more input for this specific procedure
1. Do query/mutation etc

**Reply by gino** [reply]
Thanks a lot! That does work and totally makes sense now, didn't even know it was possible to create middlewares without using the `t.middleware` function. I appreciate it!

**Reply by jaivinwylde** [reply]
very smart monad design

**Reply by pencilcheck** [reply]
In one of my middleware I want to add new stuff to ctx or meta whatever it is to carry forward, but somehow typescript keeps saying that it cannot find those attributes

I don't know if putting everything into one file is the solution for it to infer it's type as it seems to be suggested here. I would like to see a version where things can be separated into separate files instead. I have it now separated and trpc keeps complaining about unknown attributes which means it couldn't infer types

I'm also not sure if the solution to constantly initTRPC with new Context type is the solution either. Another solution to put all possible values into Context type in the beginning initTRPC is also bad design in my opinion, I really don't want to put everything and hard code in the beginning, not scalable nor making sense as sometimes I just want context to change depending on the middleware chained

**Reply by arily** [reply]
> See https://trpc.io/docs/v10/procedures#multiple-input-parsers on how you can combine this to make a reusable procedure that you can use to check that
> 
> 1. User is authorized
> 2. It's a "room procedure"
> 3. The user has access to this specific "room procedure"
> 4. Add more input for this specific procedure
> 5. Do query/mutation etc

is the link dead? It seems #multiple-input-parsers no longer exists.

**Reply by KATT** [reply]
Fixed the link now. Should be deployed within a few minutes.

**Reply by GLObus303** [reply]
I got confused by this example code -> https://trpc.io/docs/server/context#example-code. Which DOES wrap it in the middleware and also says `// Infers the `session` as non-nullable`

Is the example code about something else?

**Comment by arily** [comment]
![image](https://user-images.githubusercontent.com/10007589/202178041-5948e486-7c1a-473b-8f10-39a2f965fffc.png)

my swapped context won't be infered to the next middleware, it still thinks ctx is the original Context. 
could you please check if I did anything wrong?

**Reply by KATT** [reply]
See the original answer. Same answer applies here.

**Reply by arily** [reply]
AH I see. I shouldn't wrap the handler with the middleware(). Thank you so much!
